src/spend/secret.rs

```diff
@@ -0,0 +1,144 @@
+use super::public::PublicKey;
+use super::stealth::StealthAddress;
+
+use crate::sponge;
+use crate::{JubJubExtended, JubJubScalar, ViewKey};
+
+use dusk_jubjub::GENERATOR;
+
+use std::fmt;
+use std::ops::Mul;
+
+use rand::rngs::StdRng;
+use rand::SeedableRng;
+use rand::{CryptoRng, RngCore};
+use sha2::{Digest, Sha256};
+
+/// Secret pair of `a` and `b`
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct SecretKey {
+    a: JubJubScalar,
+    b: JubJubScalar,
+}
+
+impl Default for SecretKey {
+    fn default() -> Self {
+        SecretKey {
+            a: JubJubScalar::random(&mut rand::thread_rng()),
+            b: JubJubScalar::random(&mut rand::thread_rng()),
+        }
+    }
+}
+
+impl SecretKey {
+    /// This method is used to construct a new `SecretKey` from the given secret
+    /// pair of `a` and `b`.
+    pub fn new(a: JubJubScalar, b: JubJubScalar) -> Self {
+        Self { a, b }
+    }
+
+    /// Gets `a`
+    pub fn a(&self) -> &JubJubScalar {
+        &self.a
+    }
+
+    /// Gets `b`
+    pub fn b(&self) -> &JubJubScalar {
+        &self.b
+    }
+
+    /// Generate a `sk_r = H(a · R) + b`
+    pub fn sk_r(&self, sa: &StealthAddress) -> JubJubScalar {
+        let aR = sa.R() * self.a;
+        let aR = sponge::hash(&aR);
+        aR + self.b
+    }
+
+    /// Deterministically create a new [`SecretKey`] from a random number
+    /// generator
+    pub fn from_rng<R: RngCore + CryptoRng>(rng: &mut R) -> Self {
+        let a = JubJubScalar::random(rng);
+        let b = JubJubScalar::random(rng);
+
+        SecretKey::new(a, b)
+    }
+
+    /// Derive the secret to deterministically construct a [`PublicKey`]
+    pub fn public_key(&self) -> PublicKey {
+        let A = JubJubExtended::from(GENERATOR).mul(&self.a);
```

Comment:
Use `jubjub::GENERATOR_EXTENDED` instead (as in the occurrences below)

Reply:
This is actually a leftover. It should be `GENERATOR.to_niels()`.
Is the `jubjub::GENERATOR_EXTENDED` a recent impl? I didn't had it in jubjub when I started the code. I will move the code to that then.